### ISSUE TITLE
[Feature Request] add constant with IP address

### ISSUE THREAD

**Issue by avigeilpro** [openshwprojects/OpenBK7231T_App#914]
Hi, can anyone add a constant with ip address ?
I would like that all my devices publish their ip address on a unique MQTT topic with the publish command and 1 as third argument, but for this i need the ip address in a constant, something like this :
`publish OpenBeken/IPList $IPAddr 1`

thanks for any help

**Comment by openshwprojects** [comment]
Good idea. I will try to add this feature for you tomorrow in the morning, however first I must check if we support text variable expansion.

**Comment by avigeilpro** [comment]
great, thank you very much, I didn't expect such a quick reaction.
I need this because i write an auto-update program with Node-Red in Home Assistant (home assistant is in http in local wich is perfect to deliver the rbl for OTA update). Just wondering : where should i post the result ? Is there a forum section for this ?

**Comment by openshwprojects** [comment]
I've pushed a draft of $IP costant, it should work, can you check?
https://github.com/openshwprojects/OpenBK7231T_App/commit/bc730699a370330696fa1d42ca2537b581a8b4d0

I will have to rewrite it later, because the core constant system didn't expect STRING constants, only floats.

Very good idea with that tool. You can post it as a tutorial here:
https://www.elektroda.com/rtvforum/forum517.html
And maybe apply for a free gift:
https://www.elektroda.com/rtvforum/topic3950844.html

**Comment by avigeilpro** [comment]
it doesn't work, i did the update, and when i try echo $IP (or $ShortName or $Name as i saw you added it), it doesn't work:
`Info:CMD:$IP`

**Comment by avigeilpro** [comment]
Hi, that first part work, but is there a way to concatenate with some string ?
example : we actually can do this :
`publish OpenBeken/DeviceList $IP 1`
or
`publish $IP whatIwant 1`

but is there a way to do this :
`publish OpenBeken/Updater/$IP ON 1`

I would need something like this that we can register and unregister devices from update, and secondly that if the server reboot, it will get all messages from all devices instead of just the last device.

thanks

**Comment by avigeilpro** [comment]
Hi, i have write a piece of code. I have tested it out of context but it seems to work:
with this you can put topic and value like this :
`publish OpenBeken/Updater/$ShortName ${IP}_$uptime 1`
You can use either $IP or ${IP}, but if you have some text after the constant name, you need to use {}

here the code : 

`static char g_argsExpanded[MAX_ARGS][40];`

`const char *Tokenizer_GetArg(int i) {
	const char *s;
    char tokLine[sizeof(g_argsExpanded[i])];
    char Templine[sizeof(g_argsExpanded[i])];
    char convert[10];
    
	if(i >= g_numArgs)
		return 0;

	s = g_args[i];

#if 0
	if(g_bAllowExpand && s[0] == '$' && s[1] == 'C' && s[2] == 'H') {
		int channelIndex;
		int value;

		channelIndex = atoi(s+3);
		value = CHANNEL_Get(channelIndex);
		
		sprintf(g_argsExpanded[i],"%i",value);

		return g_argsExpanded[i];
	}
#else
    //séparators for strtok to detect constants
	const char * separators = "${}";
	//pointer for strstr function
	char *ptrConst;
	
	//copy input string before manipulations
	strcpy_safe(g_argsExpanded[i],s,sizeof(g_argsExpanded[i]));
    strcpy_safe(tokLine,s,sizeof(tokLine));
    
    //start strtok
    char *strToken = strtok(tokLine,separators);
    while ( strToken != NULL ) {
        //build tconst with ${<token>} and try find it
        char tconst[20] = "${";
        strcat(tconst,strToken);
        strcat(tconst,"}");
        ptrConst=strstr(g_argsExpanded[i],tconst);
        if (ptrConst==NULL){
            // we didn't find ${<token>} so we try with $<token>
            strcpy(tconst,"$");
            strcat(tconst,strToken);
            ptrConst=strstr(g_argsExpanded[i],tconst);
        }
        // if we found ${<token>} or $<token> it means we found a constant
        if (ptrConst!=NULL){
            //put 0 on the start of the constant to copy the left part of the input string
            ptrConst[0]=0;
            strcpy_safe(Templine,g_argsExpanded[i],sizeof(Templine));
            //analyse the constant found to replace it with it's value/string and concat it with the left part of the input string
    		if (!strcmp(tconst, "${IP}") || !strcmp(tconst, "$IP")) {
    			strcat_safe(Templine, HAL_GetMyIPString(), sizeof(Templine));
            } else if (!strcmp(tconst, "${ShortName}") || !strcmp(tconst, "$ShortName")) {
            	strcat_safe(Templine, CFG_GetShortDeviceName(), sizeof(Templine));
            } else if (!strcmp(tconst, "${Name}") || !strcmp(tconst, "$Name")) {
            	strcat_safe(Templine, CFG_GetDeviceName(), sizeof(Templine));
            }
            else {
            	float f;
            	int iValue;
            	CMD_ExpandConstant(tconst, 0, &f);
            	printf("float %.6f\n",f);
            	iValue = f;
            	sprintf(convert, "%i", iValue);
            	strcat_safe(Templine, convert, sizeof(Templine));
            }
            //concat with the right part, after the constant
            strcat_safe(Templine,ptrConst+strlen(tconst),sizeof(Templine));
            //update the input string with the replaced constant
            strcpy_safe(g_argsExpanded[i],Templine,sizeof(g_argsExpanded[i]));
        }
        //look for next token
        strToken = strtok ( NULL, separators);
        
    }

    return g_argsExpanded[i];
    
#endif

	//return g_args[i];
}`

**Comment by openshwprojects** [comment]
Good job, but I must think about it. It is important change and I don't want to mess up syntax.

I will also probably merge it myself with numerical constants later.

**Comment by openshwprojects** [comment]
I am afraid it doesnt pass the self tests at the moment:
![image](https://github.com/openshwprojects/OpenBK7231T_App/assets/85486843/29e179a6-6dd8-4f71-80c5-fc09b3df8205)

**Comment by openshwprojects** [comment]
This code also breaks all longer arguments, I had to disable it totally for now

**Comment by avigeilpro** [comment]
I will work on this and will try come back with something more functionnal.

**Comment by avigeilpro** [comment]
Hi, i would like to do the self tests. If i understand it can be done only with visual studio. I read that it need the 2017 version, but i only have the 2022. Can I do it with this version or not ? if not where can i find the 2017 version ?
I've made some changes to the code and would like to do all the tests before a new commit and PR.

**Comment by avigeilpro** [comment]
I was finally able to do the self tests, solved the problem of longer arguments, new PR started.

**Comment by openshwprojects** [comment]
Hmm sorry, 2017 is still available but I think you can upgrade to 2022 as well.

Can I firfst add more self tests, just to be sure? Or maybe I will merge first and then add, hmm.
